python/ray/serve/backend_state.py

```diff
@@ -604,22 +573,72 @@ def __init__(self, name: str, controller_name: str, detached: bool,
+        running in. Might be lost or re-constructed upon ray cluster failure.
+        """
+        return (self._rollback_info, self._curr_goal,
                 self._prev_startup_warning,
                 self._replica_constructor_retry_counter, self._replicas)
 
-    def recover_from_checkpoint_data(self, data):
-        (self._target_info, self._rollback_info, self._target_replicas,
-         self._curr_goal, self._target_version, self._prev_startup_warning,
-         self._replica_constructor_retry_counter, self._replicas) = data
+    def get_checkpoint_data(self):
+        return (self.get_target_state_checkpoint_data(),
+                self.get_current_state_checkpoint_data())
+
+    def recover_target_state_from_checkpoint(self, target_state_checkpoint):
+        logger.info("Recovering target state for deployment "
+                    f"{self._name} from checkpoint..")
+        (self._target_info, self._target_replicas,
+         self._target_version) = target_state_checkpoint
+
+    def recover_current_state_from_checkpoint(self, current_state_checkpoint):
+        logger.info("Recovering current state for deployment "
+                    f"{self._name} from checkpoint..")
+        (self._rollback_info, self._curr_goal, self._prev_startup_warning,
+         self._replica_constructor_retry_counter,
+         self._replicas) = current_state_checkpoint
+
         if self._curr_goal is not None:
             self._goal_manager.create_goal(self._curr_goal)
 
         self._notify_backend_configs_changed()
         self._notify_replica_handles_changed()
 
+    def recover_cur_state_from_replica_actor_names(
+            self, replica_actor_names: List[str]):
+        assert (
+            self._target_info is not None and self._target_replicas != -1
+            and self._target_version is not None), (
+                "Target state should be recovered successfully first before "
+                "recovering current state from replica actor names.")
+
+        logger.info("Recovering current state for deployment "
+                    f"{self._name} from {len(replica_actor_names)} actors in "
+                    "current ray cluster..")
+        # All current states use default value, only attach running replicas.
+        for replica_actor_name in replica_actor_names:
+            replica_name: ReplicaName = ReplicaName.from_str(
+                replica_actor_name)
+            new_backend_replica = BackendReplica(
+                self._controller_name, self._detached,
+                replica_name.replica_tag, replica_name.deployment_tag,
+                self._target_version)
+            new_backend_replica.start_or_update(self._target_info,
+                                                self._target_version)
+            self._replicas.add(ReplicaState.STARTING_OR_UPDATING,
+                               new_backend_replica)
+
+        self._notify_backend_configs_changed()
+        self._notify_replica_handles_changed()
```

Comment:
We should be delaying the replica_handles change to avoid thrashing when the controller starts up.

Reply:
This will halt all traffic in the cluster.